File: stacks/rails/app/models/hospitalization.rb
# 入院。spec/screens.md #18。ケア記録は追加した順（時系列）で表示する。並べ替えは無い。
class Hospitalization < ApplicationRecord
  belongs_to :patient
  has_many :care_records, -> { order(:recorded_at, :id) }, dependent: :destroy

  validates :admitted_on, presence: true
  validates :room, presence: true

  def open?
    discharged_on.nil?
  end

  scope :open_on, ->(date) {
    where(discharged_on: nil).or(where("discharged_on >= ?", date)).where("admitted_on <= ?", date)
  }
end

File: stacks/laravel/app/Models/Hospitalization.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Hospitalization extends Model
{
    protected $fillable = ['patient_id', 'admitted_on', 'discharged_on', 'room'];

    protected $casts = ['admitted_on' => 'date', 'discharged_on' => 'date'];

    public function patient(): BelongsTo
    {
        return $this->belongsTo(Patient::class);
    }

    public function careRecords(): HasMany
    {
        return $this->hasMany(CareRecord::class)->orderBy('recorded_at');
    }

    public function isOngoing(): bool
    {
        return $this->discharged_on === null;
    }
}
